src/aarch64/ffi.c

```diff
@@ -927,6 +997,29 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
 #endif
     }
 
+#ifdef __CHERI_PURE_CAPABILITY__
+  /*
+   * Narrow bounds of c9 to on-stack arguments excluding argument structure
+   * objects passed by value in the function call which were replaced by
+   * pointers to the objects in the on-stack arguments.
+   *
+   * We can use precise bounds for the stack capability because it is already
+   * aligned to a larger size than its effective bounds.
```

Comment:
But nsaa technically could be an unrepresentable length in absurd extreme cases even if the base is sufficiently aligned.

Reply:
Could you tell what would be an example of that, and if the assertions that are below this comment don't help with that?

Reply:
My point is you can round up the length to get a representable subset of the allocation, but the length itself might not be representable without rounding. To use smaller numbers, say state.nsaa is 1000 because of how many arguments you had, but that needs rounding up to 1016 to be representable. The assertions mean that there is a representable prefix of the stack_length-long region, but you can't bounds_set_exact here in the case that state.nsaa needs rounding up to be representable (which will still be <= stack_length).

Reply:
Yes, that looks fine now, thanks